File: app/src/main/java/io/github/erikcaffrey/kata_dagger2_mariokart/data/CharacterFakeDataSource.java
package io.github.erikcaffrey.kata_dagger2_mariokart.data;

import io.github.erikcaffrey.kata_dagger2_mariokart.R;
import java.util.LinkedList;
import java.util.List;

import io.github.erikcaffrey.kata_dagger2_mariokart.domain.model.Abilities;
import io.github.erikcaffrey.kata_dagger2_mariokart.domain.model.Character;

public class CharacterFakeDataSource {

  private final List<Character> characters;

  public CharacterFakeDataSource() {
    this.characters = new LinkedList<>();
  }

  List<Character> provideCharacterList() {
    addCharacter(getMarioBros());
    addCharacter(getLuigi());
    addCharacter(getDaisy());
    addCharacter(getDkong());
    addCharacter(getToad());
    addCharacter(getYoshi());
    addCharacter(getBowser());
    addCharacter(getWario());
    return this.characters;
  }

  private Character getMarioBros() {

    final int accelerate = 100;
    final int steer = 95;
    final int brake = 83;
    final int reverse = 87;
    final int lookBehind = 100;
    final int drift = 100;

    final String name = "Mario Bros";
    final int photo = R.drawable.ic_mario;
    final int cover = R.mipmap.mario;
    final String description = "The plumber star of the series. His kart is average across the board. When Mario is controlled by the computer, his main item used is the Star.";
    final Abilities abilities = Abilities.Builder.create()
        .setAccelerate(accelerate)
        .setBrake(steer)
        .setDrift(brake)
        .setLookBehind(reverse)
        .setReverse(lookBehind)
        .setSteer(drift)
        .build();

    return new Character(name, photo, cover, description, abilities);
  }

  private Character getLuigi() {

    final int accelerate = 97;
    final int steer = 95;
    final int brake = 73;
    final int reverse = 97;
    final int lookBehind = 87;
    final int drift = 88;

    final String name = "Luigi";
    final int photo = R.drawable.ic_luigi;
    final int cover = R.mipmap.luigi;
    final String description = "Similar to his brother, Mario, his kart also have average status across the board. Also like Mario, when computer-controlled, his main item is the Star.";
    final Abilities abilities = Abilities.Builder.create()
        .setAccelerate(accelerate)
        .setBrake(steer)
        .setDrift(brake)
        .setLookBehind(reverse)
        .setReverse(lookBehind)
        .setSteer(drift)
        .build();

    return new Character(name, photo, cover, description, abilities);
  }

  private Character getDaisy() {

    final int accelerate = 88;
    final int steer = 95;
    final int brake = 85;
    final int reverse = 97;
    final int lookBehind = 89;
    final int drift = 82;

    final String name = "Daisy";
    final int photo = R.drawable.ic_daisy;
    final int cover = R.mipmap.daisy;
    final String description = "Princess Daisy is a “tomboy.” I’m not sure how this boyishness manifests itself in the character design—her choice of an unflattering yellow dress, maybe? But this “tomboy” stuff is euphemistic nonsense, anyway. I imagine Daisy is a closeted lesbian, plucked from the socially liberal dream world of Super Mario Land’s Sarasaland and forced to mold her identity to the Mushroom Kingdom’s rigid gender roles.";
    final Abilities abilities = Abilities.Builder.create()
        .setAccelerate(accelerate)
        .setBrake(steer)
        .setDrift(brake)
        .setLookBehind(reverse)
        .setReverse(lookBehind)
        .setSteer(drift)
        .build();

    return new Character(name, photo, cover, description, abilities);
  }

  private Character getToad() {

    final int accelerate = 79;
    final int steer = 98;
    final int brake = 86;
    final int reverse = 89;
    final int lookBehind = 90;
    final int drift = 93;

    final String name = "Toad";
    final int photo = R.drawable.ic_toad;
    final int cover = R.mipmap.toad;
    final String description = "Below par speed-wise. His kart turns the best, but often loses control when hit by larger opponents. Like Princess Peach, he'll try throwing mushrooms in your way when he's controlled by the computer.";
    final Abilities abilities = Abilities.Builder.create()
        .setAccelerate(accelerate)
        .setBrake(steer)
        .setDrift(brake)
        .setLookBehind(reverse)
        .setReverse(lookBehind)
        .setSteer(drift)
        .build();

    return new Character(name, photo, cover, description, abilities);
  }

  private Character getYoshi() {

    final int accelerate = 89;
    final int steer = 98;
    final int brake = 86;
    final int reverse = 82;
    final int lookBehind = 96;
    final int drift = 88;

    final String name = "Yoshi";
    final int photo = R.drawable.ic_yoshi;
    final int cover = R.mipmap.yoshi;
    final String description = "Yoshi is the green dinosaur-like creature from Super Mario World. His kart has the highest acceleration. When playing against him, he'll try throwing eggs in your way.";
    final Abilities abilities = Abilities.Builder.create()
        .setAccelerate(accelerate)
        .setBrake(steer)
        .setDrift(brake)
        .setLookBehind(reverse)
        .setReverse(lookBehind)
        .setSteer(drift)
        .build();

    return new Character(name, photo, cover, description, abilities);
  }

  private Character getDkong() {

    final int accelerate = 90;
    final int steer = 98;
    final int brake = 89;
    final int reverse = 97;
    final int lookBehind = 89;
    final int drift = 96;

    final String name = "Donkey Kong";
    final int photo = R.drawable.ic_dkong;
    final int cover = R.mipmap.dk;
    final String description = "The son of the famous gorilla, Donkey Kong. His kart's attributes are similar to Bowser's. His main item is bananas, which is what the computer uses when controlling him.";
    final Abilities abilities = Abilities.Builder.create()
        .setAccelerate(accelerate)
        .setBrake(steer)
        .setDrift(brake)
        .setLookBehind(reverse)
        .setReverse(lookBehind)
        .setSteer(drift)
        .build();

    return new Character(name, photo, cover, description, abilities);
  }

  private Character getWario() {

    final int accelerate = 98;
    final int steer = 92;
    final int brake = 86;
    final int reverse = 88;
    final int lookBehind = 96;
    final int drift = 89;

    final String name = "Wario";
    final int photo = R.drawable.ic_wario;
    final int cover = R.mipmap.wario;
    final String description = "Wario's dirty tricks bring him dangerously close to disqualification but somehow is always greenlit for the seasons that follow. Wario is the torrential counterpart and brother to Waluigi.";
    final Abilities abilities = Abilities.Builder.create()
        .setAccelerate(accelerate)
        .setBrake(steer)
        .setDrift(brake)
        .setLookBehind(reverse)
        .setReverse(lookBehind)
        .setSteer(drift)
        .build();

    return new Character(name, photo, cover, description, abilities);
  }

  private Character getBowser() {

    final int accelerate = 89;
    final int steer = 94;
    final int brake = 87;
    final int reverse = 82;
    final int lookBehind = 94;
    final int drift = 87;

    final String name = "Bowser";
    final int photo = R.drawable.ic_bowser;
    final int cover = R.mipmap.bowser;
    final String description = "Bowser is the king of the turtle-like creatures called Koopas. His kart is good at crashing into other karts, due to the size. He'll usually try throwing lava balls to stop you.";
    final Abilities abilities = Abilities.Builder.create()
        .setAccelerate(accelerate)
        .setBrake(steer)
        .setDrift(brake)
        .setLookBehind(reverse)
        .setReverse(lookBehind)
        .setSteer(drift)
        .build();

    return new Character(name, photo, cover, description, abilities);
  }

  private void addCharacter(Character character) {
    this.characters.add(character);
  }
}

File: app/src/main/java/io/github/erikjhordanrey/kata_dagger2/data/CharacterFakeDataSource.kt
/*
 * Copyright (C) 2017 Erik Jhordan Rey.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.erikjhordanrey.kata_dagger2.data

import io.github.erikjhordanrey.kata_dagger2.R
import io.github.erikjhordanrey.kata_dagger2.domain.model.Abilities
import io.github.erikjhordanrey.kata_dagger2.domain.model.Character

class CharacterFakeDataSource : DataSource {

    private val characters: MutableList<Character> = mutableListOf()

    private val marioBros: Character
        get() {

            val accelerate = 100
            val steer = 95
            val brake = 83
            val reverse = 87
            val lookBehind = 100
            val drift = 100

            val id = "1"
            val name = "Mario Bros"
            val photo = R.drawable.ic_mario
            val cover = R.mipmap.mario
            val description = "The plumber star of the series. His kart is average across the board. When Mario is controlled by the computer, his main item used is the Star."
            val abilities = Abilities.Builder.create()
                    .setAccelerate(accelerate)
                    .setBrake(steer)
                    .setDrift(brake)
                    .setLookBehind(reverse)
                    .setReverse(lookBehind)
                    .setSteer(drift)
                    .build()

            return Character.Builder.create()
                    .setId(id)
                    .setName(name)
                    .setPhoto(photo)
                    .setCover(cover)
                    .setDescription(description)
                    .setAbilities(abilities)
                    .build()
        }

    private val luigi: Character
        get() {

            val accelerate = 97
            val steer = 95
            val brake = 73
            val reverse = 97
            val lookBehind = 87
            val drift = 88

            val id = "2"
            val name = "Luigi"
            val photo = R.drawable.ic_luigi
            val cover = R.mipmap.luigi
            val description = "Similar to his brother, Mario, his kart also have average status across the board. Also like Mario, when computer-controlled, his main item is the Star."
            val abilities = Abilities.Builder.create()
                    .setAccelerate(accelerate)
                    .setBrake(steer)
                    .setDrift(brake)
                    .setLookBehind(reverse)
                    .setReverse(lookBehind)
                    .setSteer(drift)
                    .build()

            return Character.Builder.create()
                    .setId(id)
                    .setName(name)
                    .setPhoto(photo)
                    .setCover(cover)
                    .setDescription(description)
                    .setAbilities(abilities)
                    .build()
        }

    private val daisy: Character
        get() {

            val accelerate = 88
            val steer = 95
            val brake = 85
            val reverse = 97
            val lookBehind = 89
            val drift = 82

            val id = "3"
            val name = "Daisy"
            val photo = R.drawable.ic_daisy
            val cover = R.mipmap.daisy
            val description = "Princess Daisy is a “tomboy.” I’m not sure how this boyishness manifests itself in the character design—her choice of an unflattering yellow dress, maybe? But this “tomboy” stuff is euphemistic nonsense, anyway. I imagine Daisy is a closeted lesbian, plucked from the socially liberal dream world of Super Mario Land’s Sarasaland and forced to mold her identity to the Mushroom Kingdom’s rigid gender roles."
            val abilities = Abilities.Builder.create()
                    .setAccelerate(accelerate)
                    .setBrake(steer)
                    .setDrift(brake)
                    .setLookBehind(reverse)
                    .setReverse(lookBehind)
                    .setSteer(drift)
                    .build()

            return Character.Builder.create()
                    .setId(id)
                    .setName(name)
                    .setPhoto(photo)
                    .setCover(cover)
                    .setDescription(description)
                    .setAbilities(abilities)
                    .build()
        }

    private val toad: Character
        get() {

            val accelerate = 79
            val steer = 98
            val brake = 86
            val reverse = 89
            val lookBehind = 90
            val drift = 93

            val id = "4"
            val name = "Toad"
            val photo = R.drawable.ic_toad
            val cover = R.mipmap.toad
            val description = "Below par speed-wise. His kart turns the best, but often loses control when hit by larger opponents. Like Princess Peach, he'll try throwing mushrooms in your way when he's controlled by the computer."
            val abilities = Abilities.Builder.create()
                    .setAccelerate(accelerate)
                    .setBrake(steer)
                    .setDrift(brake)
                    .setLookBehind(reverse)
                    .setReverse(lookBehind)
                    .setSteer(drift)
                    .build()

            return Character.Builder.create()
                    .setId(id)
                    .setName(name)
                    .setPhoto(photo)
                    .setCover(cover)
                    .setDescription(description)
                    .setAbilities(abilities)
                    .build()
        }

    private val yoshi: Character
        get() {

            val accelerate = 89
            val steer = 98
            val brake = 86
            val reverse = 82
            val lookBehind = 96
            val drift = 88

            val id = "5"
            val name = "Yoshi"
            val photo = R.drawable.ic_yoshi
            val cover = R.mipmap.yoshi
            val description = "Yoshi is the green dinosaur-like creature from Super Mario World. His kart has the highest acceleration. When playing against him, he'll try throwing eggs in your way."
            val abilities = Abilities.Builder.create()
                    .setAccelerate(accelerate)
                    .setBrake(steer)
                    .setDrift(brake)
                    .setLookBehind(reverse)
                    .setReverse(lookBehind)
                    .setSteer(drift)
                    .build()

            return Character.Builder.create()
                    .setId(id)
                    .setName(name)
                    .setPhoto(photo)
                    .setCover(cover)
                    .setDescription(description)
                    .setAbilities(abilities)
                    .build()
        }

    private val dkong: Character
        get() {

            val accelerate = 90
            val steer = 98
            val brake = 89
            val reverse = 97
            val lookBehind = 89
            val drift = 96

            val id = "6"
            val name = "Donkey Kong"
            val photo = R.drawable.ic_dkong
            val cover = R.mipmap.dk
            val description = "The son of the famous gorilla, Donkey Kong. His kart's attributes are similar to Bowser's. His main item is bananas, which is what the computer uses when controlling him."
            val abilities = Abilities.Builder.create()
                    .setAccelerate(accelerate)
                    .setBrake(steer)
                    .setDrift(brake)
                    .setLookBehind(reverse)
                    .setReverse(lookBehind)
                    .setSteer(drift)
                    .build()

            return Character.Builder.create()
                    .setId(id)
                    .setName(name)
                    .setPhoto(photo)
                    .setCover(cover)
                    .setDescription(description)
                    .setAbilities(abilities)
                    .build()
        }

    private val wario: Character
        get() {

            val accelerate = 98
            val steer = 92
            val brake = 86
            val reverse = 88
            val lookBehind = 96
            val drift = 89

            val id = "7"
            val name = "Wario"
            val photo = R.drawable.ic_wario
            val cover = R.mipmap.wario
            val description = "Wario's dirty tricks bring him dangerously close to disqualification but somehow is always greenlit for the seasons that follow. Wario is the torrential counterpart and brother to Waluigi."
            val abilities = Abilities.Builder.create()
                    .setAccelerate(accelerate)
                    .setBrake(steer)
                    .setDrift(brake)
                    .setLookBehind(reverse)
                    .setReverse(lookBehind)
                    .setSteer(drift)
                    .build()

            return Character.Builder.create()
                    .setId(id)
                    .setName(name)
                    .setPhoto(photo)
                    .setCover(cover)
                    .setDescription(description)
                    .setAbilities(abilities)
                    .build()
        }

    private val bowser: Character
        get() {

            val accelerate = 89
            val steer = 94
            val brake = 87
            val reverse = 82
            val lookBehind = 94
            val drift = 87

            val id = "8"
            val name = "Bowser"
            val photo = R.drawable.ic_bowser
            val cover = R.mipmap.bowser
            val description = "Bowser is the king of the turtle-like creatures called Koopas. His kart is good at crashing into other karts, due to the size. He'll usually try throwing lava balls to stop you."
            val abilities = Abilities.Builder.create()
                    .setAccelerate(accelerate)
                    .setBrake(steer)
                    .setDrift(brake)
                    .setLookBehind(reverse)
                    .setReverse(lookBehind)
                    .setSteer(drift)
                    .build()

            return Character.Builder.create()
                    .setId(id)
                    .setName(name)
                    .setPhoto(photo)
                    .setCover(cover)
                    .setDescription(description)
                    .setAbilities(abilities)
                    .build()
        }

    override fun provideCharacterList(): List<Character> {
        addCharacter(marioBros)
        addCharacter(luigi)
        addCharacter(daisy)
        addCharacter(toad)
        addCharacter(yoshi)
        addCharacter(dkong)
        addCharacter(wario)
        addCharacter(bowser)
        return this.characters
    }

    private fun addCharacter(character: Character) {
        this.characters.add(character)
    }
}
